### PR TITLE
Upgrade to Coraza 3.0

### LINKED ISSUE #63: Upgrade to Coraza 3.0

### PR DESCRIPTION
**Pull request by zc-devs**
Closes #63
Superseedes #58 and  #60

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -1,26 +1,27 @@
-module github.com/corazawaf/coraza-spoa
+module coraza-spoa
```
Comment (jcchavezs) [inline go.mod:1]: Wonder about this change, is it necessary? I think it is better to be consisten with the other connectors and use `github.com/corazawaf/coraza-spoa`

File: internal/request.go [hunk 2]
```diff
@@ -27,6 +27,27 @@ func (s *SPOA) processRequest(msg spoe.Message) ([]spoe.Action, error) {
 		tx      types.Transaction
 	)
 	var app *application
+
+	defer func() {
+		if tx == nil || tx.ID() == "" || app == nil {
```
Comment (jcchavezs) [inline internal/request.go:32]: I wonder if the case `tx.ID() == ""` is needed.

File: internal/request.go [hunk 3]
```diff
@@ -134,10 +141,13 @@ func (s *SPOA) processRequest(msg spoe.Message) ([]spoe.Action, error) {
 				return nil, fmt.Errorf("invalid argument for http request body, []byte expected, got %v", arg.Value)
 			}
 
-			_, err := tx.RequestBodyWriter().Write(body)
+			it, _, err := tx.WriteRequestBody(body)
```
Comment (jcchavezs) [inline internal/request.go:144]: 👍 +1:

File: internal/spoa.go [hunk 4]
```diff
@@ -193,8 +185,13 @@ func New(conf *config.Config) (*SPOA, error) {
 		logger := zap.New(core)
 
 		conf := coraza.NewWAFConfig().
-			WithDirectives(strings.Join(cfg.Rules, "\n")).
-			WithErrorLogger(logError(logger))
+			WithDirectives(cfg.Directives).
+			WithErrorCallback(logError(logger))
+
+		if len(cfg.Rules) > 0 {
+			// Deprecated: this will soon be removed
```
Comment (jcchavezs) [inline internal/spoa.go:192]: Probably adding a warn log.

**Review by jcchavezs — APPROVED** [review]

File: internal/request.go (hunk 2 above)
Comment (zc-devs) [reply internal/request.go:32]: Looks like [not needed](https://github.com/corazawaf/coraza/blob/ad50864cfa00f5143fff7f6877847326fdc4b394/internal/corazawaf/waf.go#L140)

**Comment by jcchavezs** [comment]
Please fix the lint. I think best option for the offending lines is to add a nolint annotation.

File: internal/response.go [hunk 5]
```diff
@@ -91,10 +91,13 @@ func (s *SPOA) processResponse(msg spoe.Message) ([]spoe.Action, error) {
 			if !ok {
 				return nil, fmt.Errorf("invalid argument for http response body, []byte expected, got %v", arg.Value)
 			}
-			_, err := tx.ResponseBodyWriter().Write(body)
+			it, _, err := tx.WriteRequestBody(body)
```
Comment (rikatz) [inline internal/response.go:94]: @jcchavezs thanks for the project! I am looking into it to do something similar to nginx and ingress nginx :)

While copying (eh, sorry!) this part of the code, I've realized this is WriteRequestBody, but shouldn't it be WriteResponseBody as we are on the response part?

Comment (zc-devs) [reply internal/response.go:94]: Fixed in #68 ([spoa.go#L375](https://github.com/corazawaf/coraza-spoa/blob/57efefa1a3f0e9086ac00198849d97bad5c1b99a/internal/spoa.go#L375))

**Pull request merged**